Objective-C/Internal/Replicator/CBLRemoteDatabase+Internal.h

```diff
@@ -0,0 +1,120 @@
+//
+
+#import "CBLRemoteDatabase.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface ConnectedClientContext : NSObject
+@property (nonatomic) CBLRemoteDatabase* remoteDB;
+
+- (instancetype) init NS_UNAVAILABLE;
+- (instancetype) initWithRemoteDB: (CBLRemoteDatabase*)db;
+@end
+
+@implementation ConnectedClientContext
+@synthesize remoteDB=_remoteDB;
+
+- (instancetype) initWithRemoteDB: (CBLRemoteDatabase *)db {
+    self = [super init];
+    if (self) {
+        _remoteDB = db;
+    }
+    return self;
+}
+
+@end
+
+@interface ConnectedClientGetDocumentContext : ConnectedClientContext
+@property (nonatomic) void(^docGetCompletion)(CBLDocument*, NSError*);
+
+- (instancetype) init NS_UNAVAILABLE;
+- (instancetype) initWithRemoteDB: (CBLRemoteDatabase*)db
+                       completion: (void(^)(CBLDocument*, NSError*))completion;
+@end
+
+@implementation ConnectedClientGetDocumentContext
+@synthesize docGetCompletion=_docGetCompletion;
+
+
+- (instancetype) initWithRemoteDB: (CBLRemoteDatabase *)db
+                       completion: (void (^)(CBLDocument *, NSError *))completion {
+    self = [super initWithRemoteDB: db];
+    if (self) {
+        _docGetCompletion = completion;
+    }
+    return self;
+}
+
+@end
+
+@interface ConnectedClientUpdateDocumentContext : ConnectedClientContext
+@property (nonatomic) NSString* docID;
+@property (nonatomic) FLSliceResult docBody;
+@property (nonatomic) BOOL isDeleted;
+@property (nonatomic) void(^docUpdateCompletion)(CBLDocument*, NSError*);
+
+- (instancetype) init NS_UNAVAILABLE;
+
+- (instancetype) initWithRemoteDB: (CBLRemoteDatabase*)db
+                            docID: (NSString*)docID
+                          docBody: (FLSliceResult)docBody
+                       completion: (void(^)(CBLDocument*, NSError*))completion;
+
+- (instancetype) initDeletionWithRemoteDB: (CBLRemoteDatabase *)db
+                               completion: (void (^)(CBLDocument*, NSError *))completion;
+
+@end
+
+@implementation ConnectedClientUpdateDocumentContext
+
+@synthesize docUpdateCompletion=_docUpdateCompletion, docID=_docID, docBody=_docBody;
+@synthesize isDeleted=_isDeleted;
+
+- (instancetype) initWithRemoteDB: (CBLRemoteDatabase *)db
+                            docID: (NSString*)docID
+                          docBody: (FLSliceResult)docBody
+                       completion: (void (^)(CBLDocument*, NSError *))completion {
+    self = [super initWithRemoteDB: db];
+    if (self) {
+        _docID = docID;
+        _docBody = FLSliceResult_Retain(docBody);
+        _docUpdateCompletion = completion;
+    }
+    return self;
+}
+
+- (instancetype) initDeletionWithRemoteDB: (CBLRemoteDatabase *)db
+                               completion: (void (^)(CBLDocument*, NSError *))completion {
+    self = [super initWithRemoteDB: db];
+    if (self) {
+        _isDeleted=true;
+        _docUpdateCompletion = completion;
+    }
+    return self;
+}
+
+- (void) dealloc {
+    if (!_isDeleted)
```

Comment:
It is better to do `if (!_docBody)` to follow with `FLSliceResult_Release(_docBody);`